pandas/_libs/lib.pyx

```diff
@@ -2606,6 +2610,9 @@ def maybe_convert_objects(ndarray[object] objects,
         if seen.is_bool:
             # is_bool property rules out everything else
             return bools.view(np.bool_)
+        elif convert_to_nullable_dtype and seen.is_bool_or_na:
+            from pandas.core.arrays import BooleanArray
+            return BooleanArray(bools.view(np.bool_), mask)
```

Comment:
should we default to this or the pyarrow version?

Reply:
I'm fine deferring this to `'boolean'`. In my IO work context, where there's no pyarrow reader, I've been generally relying on these functions to do inference & return EA types then converting that to pyarrow after. 

Could consider changing in the future to prevent a conversion

Reply:
Agreed, this would make us inconsistent with other areas where we are doing similar things